crypto/cipher/aes_gcm_double.c

```diff
@@ -0,0 +1,686 @@
+    int inner_aad_size = RTP_HEADER_SIZE + (4 * cc);
+    if (inner_aad_size > c->aad_size) {
+        return srtp_err_status_bad_param;
+    }
+
+    /* Set the AAD for the inner transform and reset the X bit */
+    debug_print(srtp_mod_aes_gcm_double, "inner aad: %s",
+                srtp_octet_string_hex_string(c->aad, inner_aad_size));
+    err = c->inner->type->set_aad(c->inner->state, c->aad, inner_aad_size);
+    if (err != srtp_err_status_ok) {
+        return err;
+    }
+
+    /* Reset the X bit and set the AAD for the outer transform */
+    c->aad[0] |= x;
+    debug_print(srtp_mod_aes_gcm_double, "outer aad: %s",
+                srtp_octet_string_hex_string(c->aad, c->aad_size));
+    err = c->outer->type->set_aad(c->outer->state, c->aad, c->aad_size);
+    if (err != srtp_err_status_ok) {
+        return err;
+    }
+
+    /* Copy the plaintext into a working buffer */
+    // XXX(rlb@ipv.sx): This is super wasteful, but avoids a lot of
+    // of the interpretation risks of the current API
+    unsigned int pt_size = *enc_size;
+    unsigned int ct_size = *enc_size;
+    unsigned char
+        ct[GCM_DOUBLE_MAX_PLAINTEXT_LEN + GCM_DOUBLE_MAX_AUTH_TAG_LEN];
+    if (*enc_size > GCM_DOUBLE_MAX_PLAINTEXT_LEN) {
+        return srtp_err_status_bad_param;
+    }
+    memset(ct, 0, GCM_DOUBLE_MAX_PLAINTEXT_LEN + GCM_DOUBLE_MAX_AUTH_TAG_LEN);
+    memcpy(ct, buf, *enc_size);
+
+    debug_print(srtp_mod_aes_gcm_double, "inner plaintext: %s",
+                srtp_octet_string_hex_string(ct, ct_size));
+
+    /* Perform the inner encryption */
+    err = c->inner->type->encrypt(c->inner->state, ct, &ct_size);
+    if (err != srtp_err_status_ok) {
+        return err;
+    }
+
+    /* Read the inner tag into the buffer */
+    unsigned int tag_size = 0;
+    err = c->inner->type->get_tag(c->inner->state, ct + ct_size, &tag_size);
+    if (err != srtp_err_status_ok) {
+        return err;
+    }
+
+    debug_print(srtp_mod_aes_gcm_double, "inner ciphertext: %s",
+                srtp_octet_string_hex_string(ct, ct_size + tag_size));
+
+    /* Append a null byte */
+    ct_size += tag_size + 1;
+
+    debug_print(srtp_mod_aes_gcm_double, "outer plaintext: %s",
+                srtp_octet_string_hex_string(ct, ct_size));
+
+    /* Perform the outer encryption */
+    err = c->outer->type->encrypt(c->outer->state, ct, &ct_size);
+    if (err != srtp_err_status_ok) {
+        return err;
+    }
+
+    /* Read the outer tag into the buffer */
+    tag_size = 0;
+    err = c->outer->type->get_tag(c->outer->state, ct + ct_size, &tag_size);
+    if (err != srtp_err_status_ok) {
+        return err;
+    }
+
+    ct_size += tag_size;
+
+    debug_print(srtp_mod_aes_gcm_double, "outer ciphertext: %s",
+                srtp_octet_string_hex_string(ct, ct_size));
+
+    /* Copy the ciphertext and tag to their respective buffers */
+    memcpy(buf, ct, pt_size);
+
+    c->tag_size = ct_size - pt_size;
+    memcpy(c->tag, ct + pt_size, c->tag_size);
```

Comment:
This is copying the encrypted tag after the double encryption of the inner tag + the plaintext tag from the outer encryption, I think. Do we want that?  While there are two tags (inner and outer), should we not just report the outer tag value?  (I cannot recall where this is used, so maybe this makes sense.  I don't know.)